[Update] v1.0.2b

TokyoSU opened the pull request:
hi @TeslaRus,
changelog here: [Releases](https://github.com/TokyoSU/OpenTomb/releases)
ps: need to change cmakelist for weapons.cpp,
i'm not using cmake so cmakelists is not included here !

TokyoSU commented on scripts/level/tr1/LEVEL6.lua:
```diff
@@ -6,6 +6,7 @@ level_PostLoad = function()
     moveEntityLocal(player, 0, 0, 256);
     setEntityMoveType(player, MOVE_UNDERWATER);
     setEntityAnim(player, ANIM_TYPE_BASE, 108, 0);
+    rewriteName(player, ITEM_PUZZLE_1, ITEM_NAME_TR1_PUZZLE_GOLD_BAR);
```
not found the iron bar :x

TeslaRus replied:
iron bar id == 100

TokyoSU replied:
thanks :)

TokyoSU replied:
#added

TeslaRus commented:
okay... I need some time to review it ;-)

TokyoSU commented on src/gui/gui.cpp:
```diff
@@ -543,18 +550,20 @@ bool Gui_LoadScreenAssignPic(const char* pic_name)
     size_t base_len = strlen(Engine_GetBasePath());
     size_t buf_len = pic_len + base_len + 5;
-    char image_name_buf[buf_len];
+    char *image_name_buf = new char[buf_len];
```
cause error in windows.

TokyoSU commented on src/gui/gui_inventory.h:
```diff
@@ -115,17 +117,20 @@ enum inventoryState
     char                        m_label_item_name_text[GUI_LINE_DEFAULTSIZE];
 
+    // get item_id by view or selected !
+    uint32_t getItemIdActualView();
```
now used, sorry for missed using at the preview update.

TokyoSU commented:
the #c3f200e / #3149b84 and #7604359 are fail, sorry for it :x

RICCIARDI-Adrien commented on src/weapons.h:
```diff
@@ -0,0 +1,382 @@
+    int echo;
+    int reload_1;     // classic reload
+    int reload_2;     // grenadegun reload lock
+    int damage = NULL;
```
The 3 preceding fields are not pointers, they should not be initialized using NULL (this triggers a warning when building with gcc).
```suggestion
    int damage = 0;
```

TokyoSU replied:
#fixed

RICCIARDI-Adrien commented on src/weapons.h:
```diff
@@ -0,0 +1,382 @@
+    int reload_1;     // classic reload
+    int reload_2;     // grenadegun reload lock
+    int damage = NULL;
+    int damage_water = NULL;
```
```suggestion
    int damage_water = 0;
```

TokyoSU replied:
#fixed

RICCIARDI-Adrien commented on src/weapons.h:
```diff
@@ -0,0 +1,382 @@
+    int reload_2;     // grenadegun reload lock
+    int damage = NULL;
+    int damage_water = NULL;
+    int damage_explosion = NULL;
```
```suggestion
    int damage_explosion = 0;
```

TokyoSU replied:
#fixed thanks for reply

RICCIARDI-Adrien commented on src/weapons.h:
```diff
@@ -0,0 +1,382 @@
+    SND_HARPOON_RELOAD_LAND = TR_AUDIO_SOUND_RELOADHARPOON_G,
+    SND_HARPOON_RELOAD_WATER = TR_AUDIO_SOUND_RELOADHARPOON_W,
+    SND_GRENADEGUN = TR_AUDIO_SOUND_SHOTGRENADEGUN,
+    SND_TR2_3_GRENADEGUN_RELOAD = TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN,
```
I did not find `TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN` declaration in the code (is it commited ?) :
```
In file included from /home/ar/Documents/OpenTombTokyoSU/src/gui/gui_inventory.cpp:30:
/home/ar/Documents/OpenTombTokyoSU/src/gui/../weapons.h:221:35: error: ‘TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN’ was not declared in this scope
```

TokyoSU replied:
#fixed missing audio.h

RICCIARDI-Adrien commented on src/weapons.h:
```diff
@@ -0,0 +1,382 @@
+    SND_HARPOON_RELOAD_WATER = TR_AUDIO_SOUND_RELOADHARPOON_W,
+    SND_GRENADEGUN = TR_AUDIO_SOUND_SHOTGRENADEGUN,
+    SND_TR2_3_GRENADEGUN_RELOAD = TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN,
+    SND_TR2_3_GRENADEGUN_LOCK = TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN_LOCK,
```
```
/home/ar/Documents/OpenTombTokyoSU/src/gui/../weapons.h:222:33: error: ‘TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN_LOCK’ was not declared in this scope
```

RICCIARDI-Adrien commented on src/weapons.h:
```diff
@@ -0,0 +1,382 @@
+    SND_GRENADEGUN = TR_AUDIO_SOUND_SHOTGRENADEGUN,
+    SND_TR2_3_GRENADEGUN_RELOAD = TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN,
+    SND_TR2_3_GRENADEGUN_LOCK = TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN_LOCK,
+    SND_TR4_C_GRENADEGUN_RELOAD = TR_AUDIO_SOUND_TR4C_RELOADGRENADEGUN,
```
```
/home/ar/Documents/OpenTombTokyoSU/src/gui/../weapons.h:223:35: error: ‘TR_AUDIO_SOUND_TR4C_RELOADGRENADEGUN’ was not declared in this scope
```

RICCIARDI-Adrien commented on src/weapons.h:
```diff
@@ -0,0 +1,382 @@
+    SND_TR2_3_GRENADEGUN_RELOAD = TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN,
+    SND_TR2_3_GRENADEGUN_LOCK = TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN_LOCK,
+    SND_TR4_C_GRENADEGUN_RELOAD = TR_AUDIO_SOUND_TR4C_RELOADGRENADEGUN,
+    SND_TR4_C_GRENADEGUN_LOCK = TR_AUDIO_SOUND_TR4C_RELOADGRENADEGUN_LOCK,
```
```
/home/ar/Documents/OpenTombTokyoSU/src/gui/../weapons.h:224:33: error: ‘TR_AUDIO_SOUND_TR4C_RELOADGRENADEGUN_LOCK’ was not declared in this scope
```

TokyoSU replied:
#fixed missing audio.h

RICCIARDI-Adrien commented on src/gui/gui.cpp:
```diff
@@ -30,6 +30,7 @@
 #include "gui.h"
 #include "gui_menu.h"
 #include "gui_inventory.h"
+#include "inventory.h"
```
```
/home/ar/Documents/OpenTombTokyoSU/src/gui/gui.cpp:33:10: fatal error: inventory.h: Aucun fichier ou dossier de ce type
```
```suggestion
#include "../inventory.h"
```

TokyoSU replied:
#fixed thanks :)

TokyoSU commented:
thanks you @RICCIARDI-Adrien for reviewing, the last commit, fix all review.

RICCIARDI-Adrien commented on src/state_control/state_control_Lara.cpp:
```diff
@@ -3279,144 +3275,84 @@ int StateControl_LaraDoOneHandWeaponFrame(struct entity_s *ent, struct  ss_anima
         bool do_aim = ent->character->cmd.action;
         float target_pos[3];
-        int inc_state;
+        int inc_state = NULL;
```
```suggestion
        int inc_state = 0;
```

TokyoSU replied:
#fixed

RICCIARDI-Adrien commented:
You're welcome @TokyoSU.
Here is the patch to add `weapons.*` files to the CMake build system :
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 518dd8b6..69e4380c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -208,6 +208,8 @@ set(OPENTOMB_SRCS
     src/skeletal_model.c
     src/trigger.cpp
     src/trigger.h
+    src/weapons.cpp
+    src/weapons.h
     src/world.cpp
     src/world.h
 )
```

RICCIARDI-Adrien commented:
There is a sound problem in TR2 : uzis firing sound does not stop playing.

RICCIARDI-Adrien commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    }
+
+    // no id found
+    return NULL;
```
```suggestion
    return 0;
```
Is 0 an invalid ID (I did not dig into IDs code, it's just out of curiosity) ?

TokyoSU replied:
0 is null because 1 is for pistol #fixed

RICCIARDI-Adrien commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    weapons[11] = grenadegun;
+    weapons[12] = crossbowgun;
+    weapons[13] = grapplingun;
+    return weapons;
```
Be careful, `weapons` is a local variable. This function won't work if called.

TokyoSU replied:
#delete_and_fixed

TokyoSU commented:
i already know for uzi sound but this bug is also for m16 :x i will fix it

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    grapplingun.muzzle_pos[1] = NULL;
+    grapplingun.muzzle_pos[2] = NULL;
+    grapplingun.muzzle_orient[0] = NULL;
+    grapplingun.muzzle_orient[1] = NULL;
```
#fixed

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    grapplingun.muzzle_pos[2] = NULL;
+    grapplingun.muzzle_orient[0] = NULL;
+    grapplingun.muzzle_orient[1] = NULL;
+    grapplingun.muzzle_orient[2] = NULL;
```
#fixed

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    grapplingun.muzzle_pos[0] = NULL;
+    grapplingun.muzzle_pos[1] = NULL;
+    grapplingun.muzzle_pos[2] = NULL;
+    grapplingun.muzzle_orient[0] = NULL;
```
#fixed

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    // no muzzleflash for this ?
+    grapplingun.muzzle_pos[0] = NULL;
+    grapplingun.muzzle_pos[1] = NULL;
+    grapplingun.muzzle_pos[2] = NULL;
```
#fixed

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    grapplingun.haveGravity = false;
+    // no muzzleflash for this ?
+    grapplingun.muzzle_pos[0] = NULL;
+    grapplingun.muzzle_pos[1] = NULL;
```
#fixed

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    pistol.dealDmgAtImpact = false;
+    pistol.haveGravity = false;
+    pistol.muzzle_duration = 3.0f;
+    pistol.muzzle_pos[0] = NULL;
```
#fixed

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    pistol.haveGravity = false;
+    pistol.muzzle_duration = 3.0f;
+    pistol.muzzle_pos[0] = NULL;
+    pistol.muzzle_pos[1] = NULL;
```
#fixed

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    pistol.muzzle_duration = 3.0f;
+    pistol.muzzle_pos[0] = NULL;
+    pistol.muzzle_pos[1] = NULL;
+    pistol.muzzle_pos[2] = NULL;
```
#fixed

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    pistol.muzzle_pos[0] = NULL;
+    pistol.muzzle_pos[1] = NULL;
+    pistol.muzzle_pos[2] = NULL;
+    pistol.muzzle_orient[0] = NULL;
```
#fixed

TokyoSU commented on src/weapons.cpp:
```diff
@@ -0,0 +1,1526 @@
+    pistol.muzzle_pos[1] = NULL;
+    pistol.muzzle_pos[2] = NULL;
+    pistol.muzzle_orient[0] = NULL;
+    pistol.muzzle_orient[1] = NULL;
```
#fixed

TokyoSU commented:
> 
> 
> You're welcome @TokyoSU.
> Here is the patch to add `weapons.*` files to the CMake build system :
> 
> ```
> diff --git a/CMakeLists.txt b/CMakeLists.txt
> index 518dd8b6..69e4380c 100644
> --- a/CMakeLists.txt
> +++ b/CMakeLists.txt
> @@ -208,6 +208,8 @@ set(OPENTOMB_SRCS
>      src/skeletal_model.c
>      src/trigger.cpp
>      src/trigger.h
> +    src/weapons.cpp
> +    src/weapons.h
>      src/world.cpp
>      src/world.h
>  )
> ```

i can just push the CMakeLists ? i not familiar with git :x

TeslaRus commented:
state control and below will be reviewed later; thanks for PR ;-)
P.S. I have to debug changes before merge too - some extra time...

TokyoSU commented:
need time too, need to commit some files before merging.

TokyoSU commented:
i want to use the "ss_anim->anim_frame_flags == ANIM_FRAME_REVERSE" but it seems that not working for unknown reason.
how to use this flag ? i use it with:
"ent->character->state.weapon_ready && !ent->character->cmd.action" for more precision.
is for uzi and m16 loop fix, but i wan't to fix echo not playing.
With (-time) echo sound are working but sound are playing when starting firing. :x

TokyoSU commented:
update 1.0.2b changelog:
- fixed exit grenadegun animation (need to fix firing anim).
- fixed uzi loop sound.
- (fixed?) revolver and crossbow lasersight in inventory are now hidden. (can be unhide with new variable "itemIsEquipped" in "weapons_s".
![screen_20190220_135250](https://user-images.githubusercontent.com/27679699/53093161-f48dd800-3516-11e9-89b4-116230d3ead7.png)
/for in-game (model on lara anim) need to be changed when lasersight is equiped !.
- reverted weapons.cpp, now this class dont use define for function (but variables use define).
![shotgun_old](https://user-images.githubusercontent.com/27679699/53092078-13d73600-3514-11e9-9227-f143d7773517.PNG)
- organizing weapons.h to be more clean.
- getVersion() now have more id for checking level version.
![version](https://user-images.githubusercontent.com/27679699/53092233-83e5bc00-3514-11e9-9f64-21e0da39fa1d.PNG)
- now hp bar will only draw in tr1 to tr3 like original games !

TokyoSU commented:
flares can be considered as a weapon ?
just need to create a dynamic light and a timer to switch off the light and launch the flare.
but for pickup the flare after launched (if light still exists), i dont know how to set up :x

TokyoSU commented:
save_path need to be created if not found or the game.cpp throw warning that save can't be created !

TokyoSU commented:
i restructured state folder:
![state_new_structuration](https://user-images.githubusercontent.com/27679699/53694352-0ccee400-3dae-11e9-9fb5-19ef10916aa8.PNG)
in tr1 folder:
![tr1_state_new_structuration](https://user-images.githubusercontent.com/27679699/53694355-19ebd300-3dae-11e9-935d-ca70017ebf68.PNG)
sure i change all #import and cmakelists.txt for you but i need your approval for it because state_control folder in opentomb repo need to be cleared before/after PR is pushed.
this restructuring of state_control folder is more clean, and distributed in multiple folder and name is lowered. this commits is for an another PR.

PS: tr2 entity now have .lua files but no function for now, tiger in test:
![tr2_entity](https://user-images.githubusercontent.com/27679699/53694432-215fac00-3daf-11e9-96dd-36816c7d1e58.PNG)

TeslaRus commented:
![screen_20190304_221150](https://user-images.githubusercontent.com/11921962/53760425-e1deb000-3ed3-11e9-9086-15450e646574.png)
Target flag cleaning was lost

TeslaRus commented:
resolved manually.
work in progress.
critical bugs must be fixed.
